eng/common/TestResources/Remove-TestResources.ps1

```diff
@@ -116,6 +119,15 @@ if (!$ResourceGroupName) {
     $ResourceGroupName = "rg-$BaseName"
 }
 
+if (![string]::IsNullOrWhiteSpace($ServiceDirectory)) {
```

Comment:
This check is unnecessary. Put `[ValidateIsNotNullOrEmpty()]` on the parameter as well and just use line 123. IT's better to do most validation in PowerShell parameters because it will validate and return correct errors.

Reply:
I noticed you pushed another commit, but please make this change as well. We want PowerShell to validate as much as possible - not the script.

Reply:
The advantage of allowing an empty value here is that invocation would be more simple for the contributor... but only if they coincidentally didn't have one of these pre- scripts (in the majority of cases today you don't need to run a pre- script for resource removal). 

If we do validation here we'll need to update references to `remove-test-resources.yml` in each repo before the PRs that this merge opens get merged or we'll be broken.

Reply:
If they don't want to specify it, don't specify the parameter. `ValidateNullOrEmpty` only applies if they specify the parameter. That's the norm with well-written PowerShell scripts.

Reply:
Thanks for clarifying... I think the correct syntax would take the form of: 

```
...

[ValidateNotNullOrEmpty()]
[string] $ServiceDirectory,

...

if ($ServiceDirectory) { 
  $root = [System.IO.Path]::Combine("$PSScriptRoot/../../../sdk", $ServiceDirectory) | Resolve-Path
  ...
}
```

I'll start updating the template invocations ahead of this change.

Reply:
You still need the `[Parameter()]` attribute.

Reply:
I actually think we should not put `ValidateNotNullOrEmpty` as that will be a breaking change for all the consumers of our `eng\common`. So at least initially I'd stick with what we is here right now. We can later decide to fix up all the places and then we can add this if we feel it is interesting.

The more and more dependencies we get on the stuff under `eng/common` the more we are going to have to be careful about making breaking changes as it will become very difficult to keep things working in all the places if we don't.

Reply:
It's not a breaking change. It only applies if you specify the `-ServiceDirectory`. If they didn't (it didn't exist), there's no harm. This is how PowerShell works.

Reply:
Combining it with the change to the template which pass empty by default will break pipelines until we can get all the users updated to pass something not-empty.  We can come back and fix this later once we have updated all the usages.